kagi/views/totp_devices.py

```diff
@@ -27,11 +27,33 @@ class AddTOTPDeviceView(OriginMixin, FormView):
     template_name = "kagi/totp_device.html"
     success_url = reverse_lazy("kagi:totp-devices")
 
-    def gen_key(self):
-        return os.urandom(20)
-
-    def get_otpauth_url(self, key):
-        secret = b32encode(key)
+    def get(self, request, *args: str, **kwargs):
+        # When opening the view with a GET request, we treat it as a "add new
+        # device" request. There, we create a new TOTP secret and put it into
+        # the current user's session. Upon POST, the secret is read from the
+        # session again.
+        # Once a new TOTP device was successfully added, we'll drop the secret
+        # from the session.
+        # This approach allows to re-enter the token if mistyped, while keeping
+        # the same TOTP device setup on the TOTP generator.
+        self.secret = self.gen_secret()
+        request.session[SESSION_TOTP_SECRET_KEY] = self.secret
```

Comment:
The code looks sounds, but I would like to think us of the consequences of recreating the secret for every get request. Would it hurt if we just reused an existing secret from the session? I am just worried about the error rates due to concurrent access synced between mobile and pc or similar (though I am not sure if that is really worth thinking about). On the other hand, if the secret stays constant till it is actually used we certainly want a button/link to change to a new secret. Any thoughts?

Reply:
We should certainly consider what it means, if a user would maliciously hit the endpoint over and over again. We may not exhaust the `/dev/urandom` data, but the data read would not necessarily be cryptographically safe.

Reply:
I am not sure that is a realistic attack vector. session ids are random as well and probably won't even require a login to create (depending on the app) and would suffer from the same problem?